### ISSUE TITLE
Use case

### ISSUE THREAD

**Issue by massich** [pydicom/pydicom-datasets#2]
Lets say that I have some data. [like this shared gdrive directory](https://drive.google.com/drive/folders/0B5eF8SN8T5rUZWtnT1Y0RkVWb3c?usp=sharing) containing an image folder with a set of slice images, and a rtss directory with a single file containing a volume annotation.

What are the instructions to put it in **pydicom-datasets** and how can we retrieve it afterwards to use. Shall we construct a **dicom-coockie**? 

I'm a bit confused

**Comment by vsoch** [comment]
Yes we could! Could you tell me more about how slices and volumes are used, and how they are moved around / read in pydicom?

**Comment by massich** [comment]
Well, the doctors share a set of directories like the one i shared with you guys. Some times a dicom dir with volumes, other times a dicom dir with a set of slices (like the one I shared). Then I can manage them in any manner, but at the end of the day I use a 3D volume to extract features. 

What do you suggest? I like the pipeline from nipype. You load the image (or dicomdir), you build a masker, then you load the RTSS apply the previous masker to this.

**Comment by vsoch** [comment]
Then it would make sense to have a simple function to either load a provided dataset, or one they have on hand (a directory of files). What I do in deid is to have some [provided dataset](https://github.com/pydicom/deid/blob/master/deid/data/__init__.py#L33), that are easy to find based on tags so the user can just do:

```
from deid.data import get_dataset
base = get_dataset('dicom-cookies')
```
and that actually returns the full path to the folder, and then a [recursive find and validation](https://github.com/pydicom/deid/blob/master/deid/dicom/utils.py#L46) is done to find all valid dicoms, so you would do:

```
from deid.dicom import get_flies
dicom_files = get_file(base)
```
so what might be reasonable for some dataset repo is to have a registry of "provided datasets," and each would come down to a registered submodule (its own git directory) that would then (if the user needs it) be plopped into a respective folder. Each dataset (in its own repo) could then still be served in a way that would give it its own status API. 

The user then has several choices. Given one or more datasets provided in the registry (with basically a record based on the github repo), the user could use the API endpoints to look around at the different metadata (without downloading anything). When he/she finds one that he/she likes, he/she could use the function "get_dataset" which would validate that the same matches to one provided in the registry, and then do the work on the back end to download from Github. Depending on what you think is the more likely use case, this would probably best have a command line util to do this searching / downloading, and present the user with a folder. Given doing from within python, it could return the folder or an organized list /manfiest of files (as the user would want to start programatically doing stuff over them).

Given that the user has a folder (per your example) then you would skip the "get_dataset" part and just go right into "get_files." You would again use the recursive find with validation to return a datastructure that has different file types organized. If there is some common steps / functions (eg, like in nipype it's common to build a master) you would want to do something similar.

As for where the scripts live, it seems most appropriate this could be a part of pydicom, and then this repo would be the static registry with an easy to grab manifest of what is currently available. That would also support other researchers not in the pydicom organization to PR and add links to their datasets.

**Comment by vsoch** [comment]
and I think we would want an easy to use script that can take in an input directory, and spit out a dataset like dicom cookies, and we would only ask the user to provide some simple metadata like a name (or derive based on the folder). We could probably provide that as a script / set of containers that are served in this repo, with some nice instructions on how to get it to work. Do you want me to give that a go with your example dataset?

**Comment by vsoch** [comment]
part of the instruction would be like:

 - run this script / container with your dataset as input
 - something about de-identification, if necessary
 - write some details in the README, give credits, etc
 - instructions on how to create a github repo
 - instructions on how to do a PR to this repo to add the dataset

and this repo would have a continuous integration test that validates the new dataset based on being able to retrieve the manifest and download images. We could do a quick check of the other manifests as well to always make sure the repo is providing datasets that do exist. What do you think?

**Comment by massich** [comment]
I like a lot the idea, specially the part of being able to version the data
and explode CI, to run stuff. But there are still things I don't really
like how containers play a part. I guess that I don't have a clear idea of
how could we split a generic workflow.

* `Cookies` tells you where to find the datasets and how to manage them,
*  `pydicom` does only the read and write, `died` does the metadata
manipulation (as for anonymizing),
* `masker` manipulates the image data so that everything has the same
references (or maybe it should be stretch up to the point of mapping
image-data to **SxF** samples features matrix),
* `viewer` to inspect stuff.

If I understood your ideas correctly, and using the example [in gdrive](
https://drive.google.com/drive/folders/0B5eF8SN8T5rUZWtnT1Y0RkVWb3c) to
drive the discussion, at some point what is needed is a 3D volume of the
phantom (from all the dicom files in IMAGES), the annotation volumes (which
are all in a single RTSS file), overlap them and visualize that the overlap
is correct.

So, in your opinion, would you convert the whole thing in a stand alone
dataset?
Lets say: `phantom delineation dataset`,  or would you rather have a
dataset called phantoms and then the example we are providing here to be
only an element of this dataset?

**Comment by massich** [comment]
I like a lot the idea, specially the part of being able to version the data
and explode CI, to run stuff. But there are still things I don't really
like how containers play a part. I guess that I don't have a clear idea of
how could we split a generic workflow.

- `Cookies` tells you where to find the datasets and how to manage them,
*  `pydicom` does only the read and write, `died` does the metadata
manipulation (as for anonymizing),
* `masker` manipulates the image data so that everything has the same
references (or maybe it should be stretch up to the point of mapping
image-data to **SxF** samples features matrix),
* `viewer` to inspect stuff.

If I understood your ideas correctly, and using the example [in gdrive](
https://drive.google.com/drive/folders/0B5eF8SN8T5rUZWtnT1Y0RkVWb3c) to
drive the discussion, at some point what is needed is a 3D volume of the
phantom (from all the dicom files in IMAGES), the annotation volumes (which
are all in a single RTSS file), overlap them and visualize that the overlap
is correct.

So, in your opinion, would you convert the whole thing in a stand alone
dataset?
Lets say: `phantom delineation dataset`,  or would you rather have a
dataset called phantoms and then the example we are providing here to be
only an element of this dataset?

**Comment by vsoch** [comment]
- there is no dependency on containers. It would simply be the case that we provide the tool to convert a data folder (folder of dicoms) into a "dataset repo" (eg dicom-cookies) using a container. This comes down to just having a Dockerfile and Singularity file in the folder, so someone could run something like:

```
  docker run -B $PWD:/data vanessa/dicom-dataset --input /path/to/dicoms
```
 - cookies is just one example dataset. It's an example of what could be the folder that you shared.
 - haha, I think you meant deid, I sure hope that no one dies :)
- oh I see, so you are thinking a separate `pydicom/masker` tool? Yeah that would be cool! Definitely go for it. And then `pydicom/viewer` we probably don't need to re-invent the wheel, given how great [https://github.com/rii-mango/Papaya](https://github.com/rii-mango/Papaya) is.

These are the steps I would do. We should start with base tasks first and move up to this more complicated task that you want to do.

1. Create the containers and script (tools) for turning a folder into a dataset. Write docs, examples, test with the folder you shared. Within that dataest, make sure enough metadata is captured to distinguish different file types / purposes for later stuff (if relevant)
2. Once the dataset generator is done, we can do it for a few datasets. Then we update this repo with functions, etc, so it acts like a "registry"
3. Then we must have a way to communicate with that registry. My suggestion is a simple function to the core (pydicom) itself to do this, as this is the most likely tool people will have start from, and the name of our organization :)
4. Once we have the registry, and a tool to work with it, then we can start to put together simple (what we can think of now like pipelines, we can call them whatever) to do the kind of task you are describing.

I have mixed feelings about your question, because my preference is for volumetric data (eg nifti) but we don't need to re-invent the wheel with nibabel/nipy suite of tools, generally. Our focus is on dicom. So I would say we would serve the individual images as a phantom-[somename] dataset, and then have operations after download for the user to combine into a volume and do other things. One thing that might be useful to talk about is if different "classifications" of dicom are needed, for example, `phantom` would be a useful keyword to have.

**Comment by massich** [comment]
> I have mixed feelings about your question, because my preference is for volumetric data (eg nifti) but we don't need to re-invent the wheel with nibabel/nipy suite of tools, generally. Our focus is on dicom.

That's the question. `nipy` is a great suit for medical imaging for neuro imaging. Nifti is great, if I could just use it. But neither in prostate nor breast medical imaging communities is accepted. They use dicom. 

Maybe some parts of `nipy` should be factorized for any medical imaging pipeline, then `pydicom` should only be input output of the of the submodule equivalent to `nibabel`, and `cookies` used to manage any type of medical register. 

I don't have a clear idea.

**Comment by vsoch** [comment]
Yes I think that's a great idea and would be easy to do! For working in "dicom space," the functions are here. But perhaps we should just make a rich set of converters (to nifti space) so users can be pointed directly to the nipy set of tools.

Just to be clear, `dicom-cookies` is only an example dataset, and we would have many of that flavor. The management of what I am calling a registry (which easily could be dicom and nifti, and in fact I would just plug something into the conversion to perhaps convert to nifti and serve both) is this repo, pydicom-datasets.

The equivalent of `nibabel` here is `pydicom` - it is the base library for working with the data.

**Comment by scaramallion** [comment]
I'm still in the planning stages for pydicom-datasets. Currently there are a couple of concepts I'm still working through:
1. How are the datasets organised within the project?
2. What's the best way for the user to access that data?
3. Do I need to use a database in the backend?

At the moment I envisage that data could be added fairly simply, by calling an `index.add_directory()` or `index.add_dataset()` method or by wrapping those functions in a Storage SCP.

Retrieving the datasets would be a bit more complex, but if you know some of the details of the datasets you've added its not too bad. I've been thinking of creating a pydicom.Dataset subclass called Query that would contain either a single element or multiple elements for lookup. An example for finding your RTStructure dataset would be:
```python
query = Query()
query.SOPClassUID = '1.2.840.10008.5.1.4.1.1.481.3' # RT Structure Set Storage
query.PatientName = 'Whatever'
# Find all datasets with matching SOP Class and Patient Name
result = index.search(query)
```
Then to find the related SOP Instances (your images):
```python
# Assuming result contains only the RTStructure we are interested in
related = index.get_referenced_instances(result[0]) # Should return the related SOP Instances
```
Or alternatively you could just use the study and series UIDs:
```python
query = Query()
query.StudyInstanceUID = '1.2.3456'
query.SeriesInstanceUID = '1.2.3456.1'
result = index.search(query)
```

But as I said, this is still the planning stage. If you have implementation ideas I'd love to hear them. One thing I do need to check is whether pydicom supports elements with the extended Query/Retrieve values (particularly for DT/TM) as then it should be possible to use pydicom-datasets as the backend for a Query/Retrieve SCP.

**Comment by vsoch** [comment]
Cool! Here are some ideas, if it helps. I would jump on implementing but I'm in a tidal wave of things to do (alas, there are no coconuts!), so I haven't had the time. 

### How are the datasets organised within the project?
We would want to give users a function to easily package data in a way that could be served from either Github or some static server. This means:

```
[ folder of images ] + [  package function / template ] --> static API to serve images
```
Given a standard format of that API, the central registry (eg, this Github repo) would simply provide a list to the bases. Adding a new base would mean generating your dataset, and then a PR to this repo to add its base. The continuous integration test would parse the base to test if the structure is as expected (as the functions that we provide to search the endpoints would expect)

### What's the best way for the user to access that data?
There are a couple of options! This repo that has the list of bases would offer a static way to browse the bases, each of which you can think of as like a manifest. If you look at [biocontainers](https://github.com/BioContainers/BioContainers.github.io) and the [site it serves](http://biocontainers.pro/registry/) it is literally just a static resource that does something similar. So - a user would go to some web portal, navigate around, and then select some subset of data. What we could easily do is generate a text file (yml or other) that describes a query that could be given to a pydicom-dataset function that then would download the user's data to some local folder to work with.

### Do I need to use a database in the backend?
For the webby part, I strongly suggest against databases. Databases mean maintaining servers (and paying for them) and risking someone hacking something and losing it. I have been a huge proponent of static sites, especially given all the great static site generators.

>> At the moment I envisage that data could be added fairly simply, by calling an index.add_directory() or index.add_dataset() method or by wrapping those functions in a Storage SCP.

Could you tell me more about "wrapping those functions in a Storage SCP?" I would advocate for a flat file, simple storage that maps cleanly from Github to filesystem to storage or other.

>> Retrieving the datasets would be a bit more complex, but if you know some of the details of the datasets you've added its not too bad. I've been thinking of creating a pydicom.Dataset subclass called Query that would contain either a single element or multiple elements for lookup. An example for finding your RTStructure dataset would be:

ohh I think I see what you are getting at - once the data is on the user's computer, how do they query it? I think there are a few options. Either we leave querying to the static APIs (that are also serving metadata, and this means we would have that extraction be part of the dataset generation) or we generate a small database for the user that is updated when he/she interacts to download data. Given the likely use case of a user on a single machine, even sqlite would be appropriate.

**Comment by scaramallion** [comment]
I think we're both interested in different use cases. My use cases (and experience) are more clinical and all resolve around having something available locally (due to patient privacy concerns), whereas yours seem more focused on having stuff online for use in research. I can certainly see the benefit to having datasets available in this manner (in my case particularly for providing test datasets for pynetdicom and pydicom and other projects), but I don't have the knowledge to implement that side of things, so I'm not sure what the requirements are or the best way to do so.

What I meant by a Storage SCP is simply having a DICOM SCP that receives datasets then stores them via pydicom-datasets so that I can run queries against them. It wouldn't be something implemented by pydicom-datasets, its just an example.

**Comment by vsoch** [comment]
oh yes, I totally understand! So if you look at [sendit](https://www.github.com/pydicom/sendit) this could be tweaked to achieve what you are talking about. Right now it has an inotify (a system watcher) that will sniff new folders (from a C-MOVE) and then import them into a database, and automatically put them into different queues for de-identification, and then (ultimately) actually moving to some other storage. So I think what we would want to do is integrate pynetdicom into a clone of this application, restructure the models in the database a bit depending on how you would want to model different objects, and then hook it up to a search interface. The application is Dockerized and also has a job queue and nice web UI so it would easily support some basic operations, and definitely viewing. I think a first (easy) goal would be to restructure / simplify send it to act in this manner. @scaramallion I had started to make simple clients for SOP/SCU in node dicom, eg [Echo SCP](https://github.com/pydicom/node-dcm/blob/master/node_dcm/providers.py#L55) that mirrored your command line scripts, so instead of a command line tool, you would do like:

```
from node_dcm.providers import Echo
echoP = Echo(name="stanford-echo",start=True)
```

and I'd like to have something like that that is simple and easy for these kind of applications! I hadn't finished because my group decided to use a command line tool to just do the C-MOVE (hence the watcher in sendit) but I'd really like this structure and intutitive import to be a thing. So my questions - could you take a look and we could talk about how to get this to work? Could you help me? I would offer to give the first go at the local database application, and then I could document how it works and show you where we would want to plug in the SCP. What do you think?

**Comment by vsoch** [comment]
hey @scaramallion ! I'm putting together `dicom-database` to achieve the goal you have in mind:

>> more clinical and all resolve around having something available locally (due to patient privacy concerns)

and specifically it would be a dockerized tool that can make it easy to move datasets around (pynetdicom3) and then make them query-able. Other than the obvious things (command line and web interface to search and query) could you give me examples of the kinds of things you want to do/ask of the application?

I see the difference now in what we were talking about - in my mind `dicom-datasets` is more of an API to register / find datasets (as in endpoints to download) and then `dicom-database` is the thing that is the clinical tool to actually do the getting and query, and further knows how to get/post dicom datasets. And you are also right that, by way of being on Github, **most** clinical data isn't going to fit into the `dicom-datasets` model. But I could still imagine having some endpoint that requires authentication to get a dataset, or even serving in some private repo (still discoverable via `dicom-datasets`) but then the person without permission would obviously get 404 on access.

**Comment by vsoch** [comment]
hey @scaramallion ! I still need to add the query and search views, but the guts of the management tool I put together today --> https://pydicom.github.io/dicom-database/interface.html I think if I'm clever I can have a console (programmatic) search even within the web browser (note, not recommended for non local use). Minimally there will be nice search for header/values in a web UI view you would expect, and then on the command line an entrypoint to query, which will be something like:

```
Image.objects.filter(header__value="blaaa")
Header.objects.filter(field="blaaaa", value="blaaa2")
```
lol, I'll do a little better example than that when I implement it, but TLDR it's just querying Django models, which always is really easy/intuitive. Stay tuned!
